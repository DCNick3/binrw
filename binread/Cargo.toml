[package]
name = "binread"
version = "0.0.0"
authors = ["jam1garner <jam1.mcleod@hotmail.com>"]
edition = "2018"
repository = "https://github.com/jam1garner/binread"
license = "MIT"
description = "A Rust crate for helping read structs from binary data using ✨macro magic✨"
readme = "README.md"
documentation = "https://docs.rs/binread"

[dependencies]
binread_derive = { git = "https://github.com/jam1garner/binread_derive" }
lazy_static = { version = "1.4" , optional=true }
#rustversion = "1.0"

[build-dependencies]
rustc_version = "0.2"

[features]
default = ["std"]
std = []
debug_template = ["std", "lazy_static", "binread_derive/debug_template"]
