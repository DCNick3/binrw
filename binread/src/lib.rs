//! A Rust crate for helping parse binary data using ✨macro magic✨
#![cfg_attr(not(feature="std"), no_std)]

#[cfg(feature = "std")]
use std as alloc;

#[cfg(not(feature = "std"))]
extern crate alloc;


#[cfg(not(feature = "std"))]
use alloc::{
    boxed::Box,
    vec::Vec,
    string::String,
};

pub mod io;
pub mod error;
pub mod endian;
pub mod file_ptr;
pub mod options;
pub mod strings;

#[cfg(feature = "std")]
#[cfg(feature = "debug_template")]
pub mod binary_template;

use core::any::{Any, TypeId};
pub use error::Error;
pub use endian::Endian;
pub use file_ptr::FilePtr;
pub use options::{ReadOptions, AfterParseOptions, Imports};
pub use strings::{NullString, NullWideString};

use io::{Read, Seek, SeekFrom};

/// Derive macro for BinRead. [Usage here](BinRead).
pub use binread_derive::BinRead;

mod binread_impls;
pub use binread_impls::*;

pub type BinResult<T> = core::result::Result<T, Error>;

pub trait BinRead: Sized {
    type Args: Any + Copy;

    fn read<R: Read + Seek>(reader: &mut R) -> BinResult<Self> {
        let args = match Self::args_default() {
            Some(args) => args,
            None => panic!("Must pass args, no args_default implemented")
        };

        Self::read_options(reader, &ReadOptions::default(), args)
    }
    
    fn read_args<R: Read + Seek>(reader: &mut R, args: Self::Args) -> BinResult<Self> {
        Self::read_options(reader, &ReadOptions::default(), args)
    }

    fn read_options<R: Read + Seek>(reader: &mut R, options: &ReadOptions, args: Self::Args) -> BinResult<Self>;

    fn after_parse<R: Read + Seek>(&mut self, _: &mut R, _: &ReadOptions, _: Self::Args, _: &AfterParseOptions) -> BinResult<()> {
        Ok(())
    }

    fn args_default() -> Option<Self::Args> {
        if TypeId::of::<Self::Args>() == TypeId::of::<()>() {
            Some(*unsafe{
                core::mem::transmute::<_, &Self::Args>(&())
            })
        } else {
            None
        }
    }
}
