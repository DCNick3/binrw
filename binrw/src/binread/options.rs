#[cfg(all(doc, not(feature = "std")))]
extern crate alloc;
use super::Endian;
#[cfg(all(doc, not(feature = "std")))]
use alloc::vec::Vec;

/// Runtime options for
/// [`BinRead::read_options()`](crate::BinRead::read_options).
#[non_exhaustive]
#[derive(Default, Clone, Copy)]
pub struct ReadOptions {
    /// The [byte order](crate::Endian) to use when reading data.
    ///
    /// Note that if a derived type uses one of the
    /// [byte order directives](crate::attribute#byte-order), this option
    /// will be overridden by the directive.
    pub endian: Endian,

    /// An absolute offset added to the [`FilePtr::ptr`](crate::FilePtr::ptr)
    /// offset before reading the pointed-to value.
    pub offset: u64,
}
