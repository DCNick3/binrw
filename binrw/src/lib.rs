#![doc = include_str!("../doc/index.md")]
#![cfg_attr(not(feature = "std"), no_std)]
#![cfg_attr(coverage_nightly, feature(no_coverage))]
#![cfg_attr(all(doc, nightly), feature(doc_cfg))]
#![warn(clippy::pedantic)]
#![warn(missing_docs)]
#![warn(rust_2018_idioms)]
// Lint: This is not beneficial for code organisation.
#![allow(clippy::module_name_repetitions)]

extern crate alloc;
// This extern crate declaration is required to use binrw_derive macros like
// NamedArgs inside binrw because the generated code references a binrw crate,
// but binrw is not a dependency of binrw so no crate with that name gets
// automatically added by cargo to the extern prelude.
extern crate self as binrw;
#[cfg(all(doc, not(feature = "std")))]
extern crate std;

#[doc(hidden)]
#[path = "private.rs"]
pub mod __private;
mod binread;
mod binwrite;
pub mod docs;
pub mod endian;
pub mod error;
pub mod file_ptr;
pub mod helpers;
pub mod io;
pub mod meta;
mod named_args;
#[doc(hidden)]
pub mod pos_value;
pub mod punctuated;
#[doc(hidden)]
pub mod strings;

#[cfg(all(doc, not(feature = "std")))]
use alloc::vec::Vec;
#[doc(inline)]
pub use {
    binread::*,
    binwrite::*,
    endian::Endian,
    error::Error,
    file_ptr::{FilePtr, FilePtr128, FilePtr16, FilePtr32, FilePtr64, FilePtr8},
    helpers::{count, until, until_eof, until_exclusive},
    named_args::*,
    pos_value::PosValue,
    strings::{NullString, NullWideString},
};

/// Derive macro generating an impl of the trait [`BinRead`].
///
/// See the [directives glossary](docs::attribute) for usage details.
pub use binrw_derive::BinRead;

/// Attribute macro used to generate an impl of the trait [`BinRead`] with
/// support for [temporary variables](docs::attribute#temp).
///
/// When using temporary variables, this attribute **must** be placed above
/// other attributes that generate code (e.g. `#[derive(Debug)]`) to ensure that
/// the deleted temporary fields aren’t visible to those macros.
///
/// See the [directives glossary](docs::attribute) for usage details.
pub use binrw_derive::binread;

/// Derive macro generating an impl of the trait [`BinWrite`].
///
/// See the [directives glossary](docs::attribute) for usage details.
pub use binrw_derive::BinWrite;

/// Attribute macro used to generate an impl of the trait [`BinWrite`] with
/// support for [temporary variables](docs::attribute#temp).
///
/// When using temporary variables, this attribute **must** be placed above
/// other attributes that generate code (e.g. `#[derive(Debug)]`) to ensure that
/// the deleted temporary fields aren’t visible to those macros.
///
/// See the [directives glossary](docs::attribute) for usage details.
pub use binrw_derive::binwrite;

/// Attribute macro used to generate an impl of both [`BinRead`] and
/// [`BinWrite`] traits with support for
/// [temporary variables](docs::attribute#temp).
///
/// When using temporary variables, this attribute **must** be placed above
/// other attributes that generate code (e.g. `#[derive(Debug)]`) to ensure that
/// the deleted temporary fields aren’t visible to those macros.
///
/// See the [directives glossary](docs::attribute) for usage details.
pub use binrw_derive::binrw;

/// Derive macro generating an impl of the trait [`NamedArgs`].
///
/// The use cases for this macro are:
///
/// 1. When manually implementing [`BinRead`] or [`BinWrite`] on a type where
///    named arguments are desired.
/// 2. When creating a
///    [custom parser or writer](docs::attribute#custom-parserswriters)
///    where named arguments are desired.
/// 3. When a named arguments type should be shared by several different types
///    (e.g. by using [`import_raw`](docs::attribute#raw-arguments) on
///    derived types, and by assigning the type to [`BinRead::Args`] or
///    [`BinWrite::Args`] in manual implementations).
///
/// # Field options
///
/// * `#[named_args(default = $expr)]`: Sets the default value for a field.
///
/// # Examples
///
/// ```
/// use binrw::{args, binread, BinRead, NamedArgs};
/// #[derive(Clone, NamedArgs)]
/// struct GlobalArgs<Inner> {
///     #[named_args(default = 1)]
///     version: i16,
///     inner: Inner,
/// }
///
/// #[binread]
/// #[br(import_raw(args: GlobalArgs<T::Args>))]
/// struct Container<T: BinRead> {
///     #[br(temp, if(args.version > 1, 16))]
///     count: u16,
///     #[br(args {
///         count: count.into(),
///         inner: args.inner
///     })]
///     items: Vec<T>,
/// }
///
/// # let mut input = binrw::io::Cursor::new(b"\x02\0\x42\0\x69\0");
/// # assert_eq!(
/// #     Container::<u16>::read_le_args(&mut input, args! { version: 2, inner: () }).unwrap().items,
/// #     vec![0x42, 0x69]
/// # );
/// ```
pub use binrw_derive::NamedArgs;

/// A specialized [`Result`] type for binrw operations.
pub type BinResult<T> = core::result::Result<T, Error>;

pub mod prelude {
    //! The binrw prelude.
    //!
    //! A collection of traits and types you’ll likely need when working with
    //! binrw and are unlikely to cause name conflicts.
    //!
    //! ```
    //! # #![allow(unused_imports)]
    //! use binrw::prelude::*;
    //! ```

    pub use crate::{
        binread, binrw, binwrite, BinRead, BinReaderExt, BinResult, BinWrite, BinWriterExt,
    };
}
