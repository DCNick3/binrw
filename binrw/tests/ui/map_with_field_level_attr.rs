use binrw::binread;

#[binread]
#[br(map = |_: u8| Test { y: 3 })]
struct Test {
    #[br(map(|x| x + 1))]
    y: u64,
}

fn main() {

}
