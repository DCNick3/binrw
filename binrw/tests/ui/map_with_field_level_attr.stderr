error: Field-level attributes cannot be used with top-level map
 --> $DIR/map_with_field_level_attr.rs:4:12
  |
4 | #[br(map = |_: u8| Test { y: 3 })]
  |            ^^^^^^^^^^^^^^^^^^^^^
